### ISSUE TITLE
Operating State PUT Requests failing

### ISSUE THREAD

**Issue by AnthonyMBonafide** [edgexfoundry/edgex-go#1296]
With the newly added model validation, PUT requests for Operating state in core-command are unexpectedly returning a `400 (Bad Request)`. The reason is the logic first attempts to unmarshal the request as an `admin.UpdateRequest`(see [code here](https://github.com/edgexfoundry/edgex-go/blob/e2651e465714af2da6dbbf37abfa36a7555c727a/internal/core/command/restDevice.go#L104)), then performs the validation. The validation for `admin.UpdateRequest`(see [code here](https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/models/adminstate.go#L51)) will fail since the `adminState` is not provided. Previous to the validation functionality being added, the expected HTTP request body would only have the `operatingState` and not `adminState` information. As I see it, we have a few options:

1. We update the logic to better reflect the fact that we want both admin and operating state specified in the update request. This means possibly updating the request model so that it is one consolidated request struct.
2. Update the logic to handle both admin and operating state updates independently so that either one or both can be submitted in the request and the data will be persisted properly.
3. Add separate URI/APIs for admin and operating state so that the APIs are more deterministic.

I personally am a fan of option 3. I do not like that one URI is possibly processing requests differently based on the request body(admin state takes precedence over operating state). I think it makes more sense to have a seperate URI/APIs for each state and the more top level URI `/api/v1/device/name/:name` accepts all the device data. The question here would be would we expect all the data to be present in the body of the request or would we just update the information that is passed and leave the unspecified data as is.

**Comment by brandonforster** [comment]
#1244 in my reading requires that a single endpoint be used, but I'll be the first to say as the author that I think my code is a good starting point and not a good solution. If we can make the the marshalling more robust I'd prefer that solution but I'm not conceptually opposed to going back to four endpoints.

**Comment by tsconn23** [comment]
There's a fourth option to consider -- It could be useful for the validation logic in the core models to return a specific error type when validation fails, similar to those found [here](https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/clients/types/errors.go). Right now we just throw the basic error from the standard library. A type check could then be used [here](https://github.com/edgexfoundry/edgex-go/blob/a7ba7128f4b2664e47889ac3597732d19c924f7e/internal/core/command/restDevice.go#L105) and since we know there could be two different states being updated, we'd then proceed to the next decode operation if the first fails to validate.

One criticism of this approach is that it could be seen as using error handling to control logic flow. But we have a situation where two discrete pieces of state can be updated on the same item via the same URL.
`PUT http://localhost:48082/api/v1/device/0eb4fcde-331d-49cd-a445-742f52f1e27a`
We should guard against this being a widespread practice, but I'd argue in this case it makes sense. And also, I do think there's some benefit to defining an `ErrContractInvalid` type. If agreed, this type should be defined in the `go-mod-core-contracts/models` package.

**Comment by AnthonyMBonafide** [comment]
Could we get around the issue of having the request structure determining the logic by encapsulating all possible data into one structure and simply not update any fields not specified? For example instead of accepting two types of request body structures:
```
{"adminState" : "LOCKED"}
```
and
```
{"operatingState" : "DISABLED"}
```

Could we instead accept just:
```
{
    "adminState" : "LOCKED",
    "operatingState" : "DISABLED"
}
```
and make both fields optional and if the data is specified it will be updated, otherwise left as is. For example, given a device with an `adminState` of `UNLOCKED` and an `operatingState` of `ENABLED` a request body of:
```
{ "adminState" : "LOCKED"}
```
would update the `adminState` to `LOCKED` and the `operatingState` will remain `ENABLED`, and the same goes if `operatingState` is specified and the `adminState` is not. This way we are accepting the same structure and logically handling the request the same without the need for processing the data differently depending on what structure is received. The downside here is that we will essentially have to inspect the incoming data to determine what fields need to be updated and update the data that is already persisted, rather than just updating the correct field with the data from the request. This can also be problematic for DTOs that have fields that are allowed to be empty/nil

**Comment by tsconn23** [comment]
If we take that approach, maybe we could put both properties into a `DeviceState` type located [here](https://github.com/edgexfoundry/go-mod-core-contracts/tree/master/requests/states)? Following from that, the `admin` and `operating` packages should be removed. Validation of the `DeviceState` would then be responsible for ensuring its properties are populated correctly and the end user would have to always specify both. It's the requirement to specify both that gives me pause because in some use case where I might want to lock or disable a device (for security or operational reasons) all I might have is the device name/id. However I admit I don't see that kind of use case in the system right now.

This will also mean the blackbox tests need to be reworked.

**Comment by AnthonyMBonafide** [comment]
Which path should we take for this issue?

**Comment by tsconn23** [comment]
@difince Is apparently smarter than us because she fixed this in PR #1333 by removing one line. Nice work, Diana! :-)

To take it one step further, I'd prefer that we not log the Error message when AdminState decoding fails, and we can solve that by type checking the returned error now that we have [ErrContractInvalid](https://github.com/edgexfoundry/go-mod-core-contracts/blob/24a36f294d12b108636e27c6184f22ea24766959/models/errors.go#L20) being returned when the contract validation fails.

**Comment by difince** [comment]
For me, these endpoints look like trying to update only one of the statuses (the switch cases  in the code and the tests written in blackbox-testing showed me that) .. but making the endpoint more flexible by enabling update of both statuses at once is also an option .... 

But if we stick with the current implementation ..yes ..we definitely need to handle correctly the error logs. 

About the name OperatingState isn't it long? What do you think about just operState (hm , State or Status - i do not know what is the difference)- it is shorter and more consistent with adminState. 

Probably I miss something or I am not that into the project yet but .. what I understand when I hear operational state - this is the real, current state of the object.  

For example, if we are talking about port, this port could be enabled (admin status) by the administrator, but if the cable was unplugged from the port the operational status is  DOWN and nobody can update it, until somebody really plug a cable into that port -> so for me it looks that only the device/port could report its operational status and then the system update it into the database and shows it to the customer/GUI

**Comment by tsconn23** [comment]
Closing. Resolved with the removal of endpoint per #1346
